### ISSUE TITLE
Healthchecks have a memory leak

### ISSUE THREAD

**Issue by nikgoodley-ibboost** [containers/podman#19161]
### Issue Description

Podman healthchecks in rootless containers seem to generate a slow kernel memory leak.

We've had multiple servers running very well bounded payloads slow before OOM-killer invocations and eventual host death over the course of days. Evidence seems to point to this being the Podman healthcheck mechanism (or, presumably, systemd but that's a bit behind the veil) generating leaks seen in the kmalloc-2k and kmalloc-4k slabs.

Eventual server death results always results because even redirecting the OOM killer to always target container's PID 1 does nothing to free up this slab memory.

Stopping or removing containers, `podman system reset` and even (purely for my understanding) killing every process under the user running the container doesn't reclaim this memory.

This may or may not be related to #18380 but stems from our same testing trying to get static resource consumption in containers with healthchecks.

### Steps to reproduce the issue

Any container running with a healthcheck seems to create the problem on at least CentOS Stream 9 and AlmaLinux 9.2.

Create a new Fedora family Linux instance, e.g. on AWS. Confirmed occuring on both CentOS Stream 9 (eu-west-2 AMI ami-0f48b12b3b1bc65b5) and latest AWS AlmaLinux base images, e.g.

`aws ec2 run-instances --image-id ami-0f48b12b3b1bc65b5 --instance-type t3a.small  ...`

Update it to latest, create a new unprivileged user and run a basic container with a health check, i.e.

```
sudo dnf -y update; sudo dnf install -y podman
sudo reboot
sudo useradd podman -m -u 4321
sudo loginctl enable-linger 4321
sudo su podman
cd ~
podman run --name=test -d --health-cmd='stat /etc/passwd || exit 1' --health-interval=1s busybox sleep 1d
```

In another session track the evolution of the kmalloc 2k and 4k slabs, e.g.

`while true; do date;  slabtop -o -s c | grep -E " kmalloc-(2|4)k"; sleep 60; done`

as usage seems to concentrate there.

I've been unable to find any way beyond rebooting to reclaim this memory, making healthchecks unsuitable for use with small nodes even with well-bounded containers.

### Describe the results you received

The kmalloc-2k and kmalloc-4k values are completely stable when the container is run without healthchecks (same steps as above but omitting the healthcheck options).

slabtop output is as follows for the above steps without healthcheck:

```
Fri Jul  7 12:41:24 PM UTC 2023
   576    554  96%    2.00K     36       16      1152K kmalloc-2k
   216    203  93%    4.00K     27        8       864K kmalloc-4k
Fri Jul  7 12:42:24 PM UTC 2023
   576    554  96%    2.00K     36       16      1152K kmalloc-2k
   216    195  90%    4.00K     27        8       864K kmalloc-4k
...
Fri Jul  7 01:19:24 PM UTC 2023
   576    550  95%    2.00K     36       16      1152K kmalloc-2k
   216    207  95%    4.00K     27        8       864K kmalloc-4k
```

When run _with_ healthchecks, the count and size monotonically increase. Stopping and deleting the container release a small fixed portion regardless of how long the container has been running, leaving the available system memory permanently reduced until reboot. The observed growth doesn't seem to be directly tied to frequency or command but it seems to almost tick-tock between slabs (perhaps this is slub_merge in action, bit beyond my usual sphere). 

```
Fri Jul  7 01:08:28 PM UTC 2023
   512    483  94%    2.00K     32       16      1024K kmalloc-2k
   200    180  90%    4.00K     25        8       800K kmalloc-4k
Fri Jul  7 01:09:28 PM UTC 2023
   512    493  96%    2.00K     32       16      1024K kmalloc-2k
   200    180  90%    4.00K     25        8       800K kmalloc-4k
...
<<< container started now >>>
...
Fri Jul  7 01:10:28 PM UTC 2023
   608    597  98%    2.00K     38       16      1216K kmalloc-2k
   216    211  97%    4.00K     27        8       864K kmalloc-4k
Fri Jul  7 01:11:28 PM UTC 2023
   640    622  97%    2.00K     40       16      1280K kmalloc-2k
   224    216  96%    4.00K     28        8       896K kmalloc-4k
...

Fri Jul  7 01:21:28 PM UTC 2023
   720    720 100%    2.00K     45       16      1440K kmalloc-2k
   256    252  98%    4.00K     32        8      1024K kmalloc-4k
Fri Jul  7 01:22:28 PM UTC 2023
   736    733  99%    2.00K     46       16      1472K kmalloc-2k
   256    252  98%    4.00K     32        8      1024K kmalloc-4k
Fri Jul  7 01:23:28 PM UTC 2023
   752    752 100%    2.00K     47       16      1504K kmalloc-2k
   256    254  99%    4.00K     32        8      1024K kmalloc-4k
Fri Jul  7 01:24:28 PM UTC 2023
   752    752 100%    2.00K     47       16      1504K kmalloc-2k
   256    253  98%    4.00K     32        8      1024K kmalloc-4k
Fri Jul  7 01:25:28 PM UTC 2023
   768    760  98%    2.00K     48       16      1536K kmalloc-2k
   264    262  99%    4.00K     33        8      1056K kmalloc-4k

```

Over the course of hours and days this allocates significant memory for small nodes, e.g. for a host running an otherwise unused container for half a day it's in the tens of MB.

```
  OBJS ACTIVE  USE OBJ SIZE  SLABS OBJ/SLAB CACHE SIZE NAME
 46832  46721  99%    1.00K   2927       16     46832K xfs_inode
202626 198469  97%    0.21K  11257       18     45028K dentry
  6960   6946  99%    4.00K    870        8     27840K kmalloc-2k
 28380  24536  86%    0.64K   2365       12     18920K inode_cache
  1916   1907  99%    8.00K    479        4     15328K kmalloc-4k
 80768  80580  99%    0.12K   2524       32     10096K kmalloc-64
...
  OBJS ACTIVE  USE OBJ SIZE  SLABS OBJ/SLAB CACHE SIZE NAME
222390 222364  99%    0.21K  12355       18     49420K dentry
 46720  46409  99%    1.00K   2920       16     46720K xfs_inode
 10152  10145  99%    4.00K   1269        8     40608K kmalloc-2k
  2716   2709  99%    8.00K    679        4     21728K kmalloc-4k
 28356  24634  86%    0.64K   2363       12     18904K inode_cache
 86432  86313  99%    0.12K   2701       32     10804K kmalloc-64
```

The concentration in these slabs is made more apparent when forcing the kernel to drop caches:

```
$ echo 2 > /proc/sys/vm/drop_caches
$ slabtop 
...
  OBJS ACTIVE  USE OBJ SIZE  SLABS OBJ/SLAB CACHE SIZE NAME
 10168  10146  99%    4.00K   1271        8     40672K kmalloc-2k
  2716   2693  99%    8.00K    679        4     21728K kmalloc-4k
 16800   8812  52%    1.00K   1050       16     16800K xfs_inode
 21108  19277  91%    0.64K   1759       12     14072K inode_cache
 17766  16692  93%    0.57K   1269       14     10152K radix_tree_node
```

For a simple curl check every 30s this has been seen to lock up hundreds of MB of memory over a few days.

### Describe the results you expected

Healthchecks should not leak memory.

### podman info output

```yaml
# podman info
host:
  arch: amd64
  buildahVersion: 1.30.0
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  - rdma
  - misc
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.7-2.el9.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.7, commit: 3c1e4548e778d692b2b7c2848d83708223a148b9'
  cpuUtilization:
    idlePercent: 97.33
    systemPercent: 1.18
    userPercent: 1.49
  cpus: 2
  databaseBackend: boltdb
  distribution:
    distribution: '"centos"'
    version: "9"
  eventLogger: journald
  hostname: ip-10-155-32-80.eu-west-2.compute.internal
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.14.0-333.el9.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 1451913216
  memTotal: 1830539264
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.8.5-1.el9.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.8.5
      commit: b6f80f766c9a89eb7b1440c0a70ab287434b17ed
      rundir: /run/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-3.el9.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.2
  swapFree: 0
  swapTotal: 0
  uptime: 0h 33m 58.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.access.redhat.com
  - registry.redhat.io
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 12815675392
  graphRootUsed: 1613398016
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 0
  runRoot: /run/containers/storage
  transientStore: false
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 4.5.1
  Built: 1686840794
  BuiltTime: Thu Jun 15 14:53:14 2023
  GitCommit: ""
  GoVersion: go1.20.4
  Os: linux
  OsArch: linux/amd64
  Version: 4.5.1
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

Running in AWS, but replicable on different Fedora-based distribution AMIs, i.e. CentOS Stream 9 and AlmaLinux 9.2

### Additional information

-

**Comment by Luap99** [comment]
I don't think this is podman bug, do you actually have a podman process around that allocates memory?
This looks more like a kernel or maybe systemd bug. Did you try to run this on a newer kernel?

**Comment by nikgoodley-ibboost** [comment]
Thanks for the quick reply. Since then I've been putting in a fair bit of after-hours soak testing on a few dozen VMs.

The TL;DR is that it seems to be solved somewhere either between Podman 4.4.1 and Podman 4.5.1 and/or kernel 5.14.0-284.11.1.el9_2.x86_64 (or 5.17.5-300.fc36.x86_64) and 5.14.0-333.el9.x86_64 so happy to close this as seems like it will eventually hit RHEL 9 mainline and it also seems dependent on containers.conf, giving us options in the interim.

Detail if anyone wants it:

I think first thing to note is I was premature fingering CentOS there, sorry - these longer tests showed that even an hour or two is not enough to show the leak definitively as when I tried to replicate for longer soak tests on CentOS Stream 9 the 2k/4k slabs are eventually reclaimed and reach some sort of equilibrium in the low MB, which is totally fine. This usage does not appear to be the same way in the latest Fedora versions, and the non-healthcheck control shows no movement at all in these slabs, but it's nevertheless not a _leak_ as it's clawed back, just different behaviour.

But it seems to also only manifest as a long-term leak when containers.conf sets at least two of these (as our base images do)

```
< log_driver = "k8s-file"
< cgroup_manager = "cgroupfs"
< events_logger = "file"
```

I'm not precisely sure which as my results were a little confused since both CentOS Stream 9 and AlmaLinux 9.2 come with a `/usr/share/containers/containers.conf` that has a commented out value - suggesting default - of `#log_driver = "k8s-file"`. The Fedora images show the default as `journald`, but `podman info | grep Driver` for both 4.4.1 and 4.5.1 show the value is `journald` if undefined. Only realised this after I'd done long tests.

However, I took out the `events_logger = 'file'` on one AL9.2 env and stopped seeing the issue. 

Triaging the differences while having to wait a few hours to be sure of behaviour, made it quite difficult to isolate. In summary though, when using the latest available AWS (eu-west-2) AMIs for each:

**No problems seen:**
Fedora 38 -  6.3.5-200.fc38.x86_64 (and Fedora Rawhide - 6.5.*) - Podman 4.5.1 - systemd 253 (253.4-1.fc38)
CentOS Stream 9 - 5.14.0-333.el9.x86_64 - Podman 4.5.1 - systemd 252 (252-8.el9)

**Problems seen (at least with the containers.conf changes above)**
Fedora 36 - 5.17.5-300.fc36.x86_64 - Podman 4.4.1 - systemd 250 (v250.3-8.fc36)
AlmaLinux 9.2 - 5.14.0-284.11.1.el9_2.x86_64 - Podman 4.4.1 - systemd 252 (252-13.el9_2)
Fedora 34 - 5.11.12-300.fc34.x86_64 - Podman 3.4.4 - behaviour quite different. No 2k usage, 4k much higher - but clearly different behaviour to the active latest RHEL codebases so not digging further.


I couldn't immediately find a change that would cover this and the few tests I did with different kernel versions didn't show differences so left it there.

**Comment by droghio** [comment]
Thank you for reporting this. I spent a while trying to figure out what was going on before seeing this issue. For the record I was able to reproduce this with Podman 4.5.1 using CentOS Stream 8 modified to use the 5.4 kernel-lt from `el-repo`:

```
$ uname -a
Linux silica 5.4.251-1.el8.elrepo.x86_64 #1 SMP Thu Jul 27 18:47:58 EDT 2023 x86_64 x86_64 x86_64 GNU/Linux
$ podman --version
podman version 4.5.1
```

I am seeing on average ~30MB of growth every day give or take. Example running your script:

```
$ uptime
 07:09:31 up 3 days,  3:37,  1 user,  load average: 0.24, 0.29, 0.23
$ head -n 3 slab.log && tail -n 3 slab.log 
Sat Aug 12 03:55:26 UTC 2023
  4699   4683  99%    4.00K    590        8     18880K kmalloc-4k             
  5744   5719  99%    2.00K    359       16     11488K kmalloc-2k             
Sat Aug 12 07:06:30 UTC 2023
  4827   4813  99%    4.00K    606        8     19392K kmalloc-4k             
  5968   5931  99%    2.00K    373       16     11936K kmalloc-2k     
```

Side note I also see tons of `systemd[XXX]: Started /usr/bin/podman healthcheck run` messages in `journalctl` but I figured that behavior itself was fine.
